### PR TITLE
LUN-480: Fix filtering with special characters

### PR DESCRIPTION
**Pull request by lunars97**
This pull request belongs to an issue on our [bugtracker](https://issues.tuerantuer.org/secure/RapidBoard.jspa?rapidView=27&view=detail).
You can find it there by looking for an issue with the key which is mentioned in the title of this pull request.
It starts with the keyword _LUN_.

### CODE REVIEW

File: src/services/helpers.ts [hunk 1]
```diff
@@ -182,27 +180,26 @@ export const calculateScore = (vocabularyItemsWithResults: VocabularyItemResult[
...
 
 export const getSortedAndFilteredVocabularyItems = (
   vocabularyItems: VocabularyItem[] | null,
   searchString: string
 ): VocabularyItem[] => {
-  const collator = new Intl.Collator('de-De', { sensitivity: 'base', usage: 'sort' })
+  const collator = new Intl.Collator('de-De', { sensitivity: 'base', usage: 'sort', localeMatcher: 'lookup' })
 
-  const normalizedSearchString = normalizeSearchString(searchString)
+  const normalizedSearchString = searchString
```
Comment (SteffiStuffel) [inline src/services/helpers.ts:199]: By changing this like that, we will have problems: e.g. when entering 
"Die Ölk" the result should be "Die Ölkanne", 
when entering "große Ölkanne", the result should be empty.

In the previous `normalizeSearchString` function this was working properly, as we there only removed the first word from the searchString if it is an Article, see:
```
  const searchStringWithoutArticle = ARTICLES.map(article => article.value).includes(
    searchString.split(' ')[0].toLowerCase()
  )
    ? searchString.substring(searchString.indexOf(' ') + 1)
    : searchString
```
If the first word is an article we remove it from the search string, if the first word is something else (not an article) we keep it in the search string.

This behaviour should not be changes by this PR.

**Review by SteffiStuffel — COMMENTED** [review]
Nice work, looks good :rocket: 

Just one small hint: Please squash the commits together into one. If you need help with this, we can have a look at it together today.

File: src/services/helpers.ts [hunk 2]
```diff
@@ -1,5 +1,4 @@
 import { AxiosResponse } from 'axios'
-import normalizeStrings from 'normalize-strings'
```
Comment (LeandraH) [inline src/services/helpers.ts:2]: It looks to me like we can then remove the package 'normalize-strings' from our code. Can you check if that is indeed the case?

Comment (lunars97) [reply src/services/helpers.ts:2]: yes, we don't use it anymore. I will remove it.

Comment (lunars97) [reply src/services/helpers.ts:2]: seems like the changes related to  package removal disappeared after committing it. There is no package on my local repository

File: .circleci/config.yml [hunk 3]
```diff
@@ -583,48 +583,7 @@ parameters:
         type: boolean
 version: 2.1
 workflows:
-    browserstack_delivery:
```
Comment (SteffiStuffel) [inline .circleci/config.yml:586]: Please undo the changes in this file.

**Review by SteffiStuffel — COMMENTED** [review]
It is not working now. If entering `ölk...` i get no result.

Edit: Also not working on main, think there is something weird with my emulator.

**Review by steffenkleinle — CHANGES_REQUESTED** [review]
I think the current solution is not perfect. It prevents users from finding words including special characters using the "normalized" character.
Example: Finding `der Abhänger` with the filter query `Abha` is not working, only the use of `Abhä` works. In my opinion both should work though.

File: src/services/helpers.ts [hunk 4]
```diff
@@ -210,7 +209,12 @@ export const getSortedAndFilteredVocabularyItems = (
   }
 
   const filteredVocabularyItems = vocabularyItems?.filter(
-    item => item.word.toLowerCase().includes(normalizedSearchString) || matchAlternative(item, normalizedSearchString)
+    item =>
+      item.word
+        .toLowerCase()
+        .normalize('NFD')
+        .replace(/[\u0300-\u036f]/g, '')
```
[the remark this answers is not in the document]
Comment (lunars97) [reply src/services/helpers.ts:216]: I just looked for solutions and stumbled upon normalize() method [https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/normalize](url) It basically decomposes combined characters into their base character and accent marks ( for ex., ä,ö, ü). The regex `/[\u0300-\u036f]/g` is used to match and remove accent marks, tildes, and other symbols that modify characters.

[the remark this answers is not in the document]
Comment (lunars97) [reply src/services/helpers.ts:216]: I thought that it might only be need on search query if someone types basic characters

[the remark this answers is not in the document]
Comment (SteffiStuffel) [reply src/services/helpers.ts:216]: Please put the regex into a variable with a speaking name, so everyone reading this code, knows what is happening here.

**Review by SteffiStuffel — COMMENTED** [review]
Please add tests for this function, it should test:
1) Ordering of the words in the list is correct. Words starting with special chars should not be placed at the end.
2) Words with ö,ä,ü are found correctly when searching for both "ö,ü,a" and "o,u,a"
3) Words with "o,a,u" are not found, when searching for "ö,ä,ü"

File: src/services/helpers.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (steffenkleinle) [reply src/services/helpers.ts:216]: > Please add tests for this function, it should test:
> 
>     1. Ordering of the words in the list is correct. Words starting with special chars should not be placed at the end.
> 
>     2. Words with ö,ä,ü are found correctly when searching for both "ö,ü,a" and "o,u,a"
> 
>     3. Words with "o,a,u" are not found, when searching for "ö,ä,ü"

Tbh I disagree with 3., if you type the normal chars (a, o, u) also the special chars (ä, ö, ü) should be found. That is also the reason why I personally would rather just keep the library...

[the remark this answers is not in the document]
Comment (steffenkleinle) [reply src/services/helpers.ts:216]: There are also other edge cases like `'`, ß...

[the remark this answers is not in the document]
Comment (lunars97) [reply src/services/helpers.ts:216]: should I revert everything back as it was with the library?

[the remark this answers is not in the document]
Comment (steffenkleinle) [reply src/services/helpers.ts:216]: I would use normalize strings on both the search string/query AND every word we are comparing. I think that is the easiest, most thorough and covers all the edge cases. If you or @SteffiStuffel would rather use a custom regex or some other way, that is completely fine for me as well :)

[the remark this answers is not in the document]
Comment (SteffiStuffel) [reply src/services/helpers.ts:216]: I also think using a library is better than custom regex :+1:

File: src/services/__tests__/helpers.spec.ts [hunk 5]
```diff
@@ -167,4 +179,148 @@ describe('helpers', () => {
...
+        {
+          id: 10,
+          type: VOCABULARY_ITEM_TYPES.lunesStandard,
+          word: 'Untergrund',
+          article: ARTICLES[1],
+          audio: '',
+          images: [{ id: 2, image: 'image' }],
+          alternatives: [],
+        },
+      ]
```
Comment (steffenkleinle) [inline src/services/__tests__/helpers.spec.ts:321]: Can't you just get those from the VocabularyItemBuilder?

**Review by steffenkleinle — APPROVED** [review]
Well done :rocket:

File: src/services/__tests__/helpers.spec.ts (hunk 5 above)
Comment (lunars97) [reply src/services/__tests__/helpers.spec.ts:321]: @steffenkleinle I wanted to compare this data with the one which was created by the **VocabularyItemBuilder**

Comment (steffenkleinle) [reply src/services/__tests__/helpers.spec.ts:321]: But basically you are just checking that `getSortedAndFilteredVocabularyItems` doesn't change the input, right? Or am I missing something?

Comment (lunars97) [reply src/services/__tests__/helpers.spec.ts:321]: @steffenkleinle I am checking the correct order of the list whether the umlaut characters are not at the very end of the list

Comment (steffenkleinle) [reply src/services/__tests__/helpers.spec.ts:321]: But you are doing
``` 
      const sortedTestData = getSortedAndFilteredVocabularyItems(sortedData, '')
      expect(sortedData).toEqual(sortedTestData)
```
which checks for equality of `sortedData` and `sortedTestData`. So actually you are just making sure that `sortedData` is still the same after calling `getSortedAndFilteredVocabularyItems` on it. Am I missing something?

**Review by steffenkleinle — APPROVED** [review]

File: src/services/__tests__/helpers.spec.ts (hunk 5 above)
Comment (lunars97) [reply src/services/__tests__/helpers.spec.ts:321]: @steffenkleinle you are right. I am comparing unsorted list with sorted

Comment (steffenkleinle) [reply src/services/__tests__/helpers.spec.ts:321]: However, the naming does not represent this, both is called "sorted", not "unsorted". And if one is sorted and one is unsorted, it should not be equal after sorting, should it?

Comment (steffenkleinle) [reply src/services/__tests__/helpers.spec.ts:321]: You should pass an unsorted list as params though then, not the already sorted list

Comment (lunars97) [reply src/services/__tests__/helpers.spec.ts:321]: @steffenkleinle you are right. Now I understand what I did here is a illogical. Thank you:)

**Review by sarahsporck — APPROVED** [review]

**Comment by sarahsporck** [comment]
Tested on simulated ios :)

File: src/services/__tests__/helpers.spec.ts (hunk 5 above)
Comment (steffenkleinle) [reply src/services/__tests__/helpers.spec.ts]: Good to merge :rocket:
